docs/codeExecution.md

```diff
@@ -0,0 +1,89 @@
+
+The primary external interface of these callers is the `call()` function, which takes five arguments:
+
+- `type`: the type of code being executed (either `question`, `course-element`, or `core-element`).
+- `directory`: the directory containing the file whose code will be executed.
+  - For questions, this is an item in a course's `questions` directory.
+  - For course elements, this is an item in a course's `elements` directory.
+  - For core elements, this is an item in PrairieLearn's `elements` directory.
+- `file`: the name of the file whose code will be executed (e.g. `server.py`)
+- `fcn`: the name of the function in `file` that will be executed (e.g. `grade` or `render`).
+- `args`: an array of JSON-encodeable arguments to the function being called.
+
+The piece of code to execute is specified by (`type`, `directory`, `file`) instead of an absolute path because the location of each file on disk may change between each type of code caller; allowing the code caller to construct the path from that information keeps the code that uses a caller agnostic to the underlying caller being used.
+
+### A full request cycle
+
+Let's walk through a typical request to view a question that requires a function in a corresponding `server.py` file to run.
+
+1. The page request is handled by `pages/studentInstanceQuestionHomework` or similar.
+2. That handler calls `getAndRenderVariant` in `lib/question` (a different function would be called if the user were submitting an answer).
+3. That function calls an internal function that calls `render` in `question-servers/freeform.js`.
+4. That function calls `getPythonCaller` in `lib/workers`. Depending on the active execution mode
+   1. If running in `container` mode, a `lib/code-caller-docker` caller will be "prepared" for the current course (which sets up necessary bind mounts) and returned.
+   2. If running in `native` mode, any available `lib/code-caller-python` caller will be returned.
+5. `call(...)` is then repeatedly invoked on the code caller with the appropriate pieces of code to be executed.
+6. Once the code caller is no longer needed during this request, `done()` is invoked on it. The forked worker is sent a `restart` message, which will cause the worker to exit and return control to the zygote. The zygote will then fork itself again, and the forked worker will wait until it receives more instructions.
+7. Page render completes and the response is sent, thus finishing the request cycle.
+
+## Operating in production
+
+When new versions of PrairieLearn are deployed, it's important to ensure that the appropriate executor image for the version being deployed is present on the machine. This ensures that PrairieLearn is able to serve traffic immediately instead of waiting for the new version to be pulled. If PrairieLearn is deployed with a `.git` directory present, the `./tools/script/pull.js` script can be used to log in to ECR and pull the appropriate version.
```

Comment:
Maybe you want to define 'ECR' here? Unless everyone knows what it is already.

Reply:
This is referring to https://aws.amazon.com/ecr/, which is already a well-established part of how we handle Docker image s in production environments.